File: 50dias/cpp/23.cpp
#include <iostream>
#include <list>
#include <sstream>
#include <cmath>

using namespace std;

void RemoveColchetesEspacosDaString(string& entrada) {
    erase(entrada, '[');
    erase(entrada, ']');
    erase(entrada, ' ');
}

list<int> ConverteListaStringsParaInteiros(string entrada) {
    RemoveColchetesEspacosDaString(entrada);

    list<int> inteirosEntrada;
    if (entrada.size() > 0) {
        stringstream sstr(entrada);
        while (sstr.good()) {
            string substring;
            getline(sstr, substring, ',');
            inteirosEntrada.push_back(stoi(substring));
        }
    }

    return inteirosEntrada;
}

string LeiaEntradaUsuario() {
    string entrada;

    cout << "O formato da entrada deve ser o seguinte:" << endl;
    cout << "[1, 2, 3, 4, 5, 6]" << endl;
    cout << "Digite uma lista de inteiros: ";
    getline(cin, entrada);

    return entrada;
}

bool EhQuadradoPerfeito(int i) {
    int raizTruncada = sqrt(i);

    return raizTruncada * raizTruncada == i;
}

void EfetuaOperacoes(list<int> inteirosEntrada) {
    int tamanhoLista = inteirosEntrada.size();
    if (tamanhoLista > 0) {
        list<int> quadradosPerfeitos;
        for (int i : inteirosEntrada) {
            if (EhQuadradoPerfeito(i))
                quadradosPerfeitos.push_back(i);
        }

        cout << "Os valores da lista que são ";
        cout << "quadrado perfeito: " << endl;
        cout << "[ ";
        for (int i : quadradosPerfeitos) {
            cout << i << " ";
        }
        cout << "]" << endl;
    }
    else
        cout << "A lista está vazia!";
}

int main()
{
    string entrada = LeiaEntradaUsuario();

    list<int> inteirosEntrada;
    inteirosEntrada = ConverteListaStringsParaInteiros(entrada);

    EfetuaOperacoes(inteirosEntrada);
    return 0;
}

File: 50dias/python/23.py
import math

def LeiaEntradaUsuario():
    print('O formato da entrada deve ser o seguinte:')
    print('[1, 2, 3, 4, ..., 30]')
    entrada = input('Digite uma lista de inteiros: ')
    return entrada

def RemoveColcheteEspacosDaString(entrada):
    entrada = entrada.replace(']', '')
    entrada = entrada.replace('[', '')
    entrada = entrada.replace(' ', '')
    return entrada

def ConverteListaStringsParaInteiros(entrada):
    entrada = RemoveColcheteEspacosDaString(entrada)

    inteirosEntrada = []
    if (len(entrada) > 0):
        entrada = entrada.split(',')

        for palavra in entrada:
            inteirosEntrada.append(int(palavra))
    return inteirosEntrada

def EhQuadradoPerfeito(i):
    raizQuadradaTruncada = int(math.sqrt(i))
    return i == raizQuadradaTruncada**2

def EfetuaOperacoes(inteiros):
    if (len(inteiros) > 0):
        quadradosPerfeitos = []
        for i in inteiros:
            if (EhQuadradoPerfeito(i)):
                quadradosPerfeitos.append(i)

        print('Os valores da lista que são quadrados perfeito:')
        print(quadradosPerfeitos)
    else:
        print('Lista está vazia!')

def main():
    entrada = LeiaEntradaUsuario();
    inteirosEntrada = ConverteListaStringsParaInteiros(entrada)
    EfetuaOperacoes(inteirosEntrada)

main()







